crates/flashblocks-archiver/src/database.rs

```diff
@@ -0,0 +1,367 @@
+use crate::types::{Flashblock, FlashblockMessage};
+use alloy_consensus::TxReceipt;
+use alloy_primitives::{Address, B256, U256};
+use alloy_rpc_types_engine::PayloadId;
+use anyhow::Result;
+use reth_optimism_primitives::OpReceipt;
+use sqlx::PgPool;
+use tracing::info;
+use uuid::Uuid;
+
+#[derive(Debug)]
+pub struct Database {
+    pool: PgPool,
+}
+
+impl Database {
+    pub fn get_pool(&self) -> &PgPool {
+        &self.pool
+    }
+
+    pub async fn new(database_url: &str, max_connections: u32) -> Result<Self> {
+        let pool = sqlx::postgres::PgPoolOptions::new()
+            .max_connections(max_connections)
+            .connect(database_url)
+            .await?;
+
+        info!("Connected to database");
+        Ok(Self { pool })
+    }
+
+    pub async fn run_migrations(&self) -> Result<()> {
+        sqlx::migrate!("./migrations").run(&self.pool).await?;
+        info!("Database migrations completed");
+        Ok(())
+    }
+
+    pub async fn get_or_create_builder(&self, url: &str, name: Option<&str>) -> Result<Uuid> {
+        let existing = sqlx::query_scalar::<_, Uuid>("SELECT id FROM builders WHERE url = $1")
+            .bind(url)
+            .fetch_optional(&self.pool)
+            .await?;
+
+        if let Some(id) = existing {
+            return Ok(id);
+        }
+
+        let id = sqlx::query_scalar::<_, Uuid>(
+            "INSERT INTO builders (url, name) VALUES ($1, $2) RETURNING id",
+        )
+        .bind(url)
+        .bind(name)
+        .fetch_one(&self.pool)
+        .await?;
+
+        info!(
+            "Created new builder: {} ({})",
+            name.unwrap_or("unnamed"),
+            url
+        );
+        Ok(id)
+    }
+
+    pub async fn store_flashblock(
+        &self,
+        builder_id: Uuid,
+        payload: &FlashblockMessage,
+    ) -> Result<Uuid> {
+        let raw_message = serde_json::to_value(payload)?;
+
+        let flashblock_id = sqlx::query_scalar::<_, Uuid>(
+            r#"
+            INSERT INTO flashblocks (
```

Comment:
Can we do the inserts as upserts (probs fine for us to tie this to Postgres)? That way when we do green/blue deploys or have multiple instances running they wont error during the time window